lib/internal/child_process.js

```diff
@@ -486,16 +499,36 @@ function setupChannel(target, channel) {
   // handlers will go through this
   target.on('internalMessage', function(message, handle) {
     // Once acknowledged - continue sending handles.
-    if (message.cmd === 'NODE_HANDLE_ACK') {
-      if (target._pendingHandle) {
-        target._pendingHandle.close();
-        target._pendingHandle = null;
+    if ((message.cmd === 'NODE_HANDLE_ACK') ||
+        (message.cmd === 'NODE_HANDLE_NACK')) {
+
+      if (message.cmd === 'NODE_HANDLE_ACK') {
+        if (target._pendingMessage) {
+          closePendingHandle(target);
+        }
+      } else {
+        // If the handle didn't reach the receiving process correctly retry up
+        // to 3 times. If it still fails, close it and print a warning.
+        if (target._pendingMessage) {
+          if (target._pendingMessage.retransmissions++ ===
+              MAX_HANDLE_RETRANSMISSIONS) {
+            closePendingHandle(target);
+            process.emitWarning('Handle didn\'t reach the receiving process ' +
+                                'correctly', 'SentHandleNotReceivedWarning');
+          }
+        }
       }
```

Comment:
This can be simplified:

```js
if (target._pendingMessage) {
  if (message.cmd === 'NODE_HANDLE_ACK') {
    closePendingHandle(target);
  } else if (target._pendingMessage.retransmissions++ ===
      MAX_HANDLE_RETRANSMISSIONS) {
    // If the handle didn't reach the receiving process correctly retry up
    // to 3 times. If it still fails, close it and print a warning.
    closePendingHandle(target);
    process.emitWarning('Handle didn\'t reach the receiving process ' +
                        'correctly', 'SentHandleNotReceivedWarning');
  }
}
```